File: ios/PurchasesHybridCommon/PurchasesHybridCommon/SKProduct+HybridAdditions.m
//
//  Created by RevenueCat.
//  Copyright © 2019 RevenueCat. All rights reserved.
//

#import "SKProduct+HybridAdditions.h"
#import "SKProductDiscount+HybridAdditions.h"

@import RevenueCat;

@implementation RCStoreProduct (RCPurchases)

- (nullable NSString *)rc_currencyCode {
    return self.currencyCode;
}

- (NSDictionary *)rc_dictionary
{
    NSMutableDictionary *d = [NSMutableDictionary dictionaryWithDictionary:@{
                        @"identifier": self.productIdentifier ?: @"",
                        @"description": self.localizedDescription ?: @"",
                        @"title": self.localizedTitle ?: @"",
                        @"price": self.price,
                        @"price_string": self.localizedPriceString,
                        @"currency_code": (self.rc_currencyCode) ? self.rc_currencyCode : [NSNull null]
                        }];
    
    d[@"intro_price"] = [NSNull null];
    d[@"intro_price_string"] = [NSNull null];
    d[@"intro_price_period"] = [NSNull null];
    d[@"intro_price_period_unit"] = [NSNull null];
    d[@"intro_price_period_number_of_units"] = [NSNull null];
    d[@"intro_price_cycles"] = [NSNull null];
    d[@"introPrice"] = [NSNull null];
    
    if (@available(iOS 11.2, tvOS 11.2, macos 10.13.2, *)) {
        if (self.introductoryDiscount) {
            d[@"intro_price"] = self.introductoryDiscount.price;
            d[@"intro_price_string"] = self.localizedIntroductoryPriceString;
            d[@"intro_price_period"] =
                [RCStoreProduct rc_normalizedSubscriptionPeriod:self.introductoryDiscount.subscriptionPeriod];
            d[@"intro_price_period_unit"] =
                [RCStoreProduct rc_normalizedSubscriptionPeriod:self.introductoryDiscount.subscriptionPeriod];
            d[@"intro_price_period_number_of_units"] = @(self.introductoryDiscount.subscriptionPeriod.unit);
            d[@"intro_price_cycles"] = @(self.introductoryDiscount.subscriptionPeriod.value);
            d[@"introPrice"] = self.introductoryDiscount.rc_dictionary;
        }
    }
    
    d[@"discounts"] = [NSNull null];
    
    if (@available(iOS 12.2, tvOS 12.2, macos 10.14.4, *)) {
        d[@"discounts"] = [NSMutableArray new];
        for (RCStoreProductDiscount* discount in self.discounts) {
            [d[@"discounts"] addObject:discount.rc_dictionary];
        }
    }
    
    return d;
}

+ (NSString *)rc_normalizedSubscriptionPeriod:(RCSubscriptionPeriod *)subscriptionPeriod API_AVAILABLE(ios(11.2), macos(10.13.2), tvos(11.2)) {
    NSString *unit;
    switch (subscriptionPeriod.unit) {
        case RCSubscriptionPeriodUnitDay:
            unit = @"D";
            break;
        case RCSubscriptionPeriodUnitWeek:
            unit = @"W";
            break;
        case RCSubscriptionPeriodUnitMonth:
            unit = @"M";
            break;
        case RCSubscriptionPeriodUnitYear:
            unit = @"Y";
            break;
    }
    return [NSString stringWithFormat:@"%@%@%@", @"P", @(subscriptionPeriod.value), unit];
}

+ (NSString *)rc_normalizedSubscriptionPeriodUnit:(RCSubscriptionPeriod *)subscriptionPeriod API_AVAILABLE(ios(11.2), macos(10.13.2), tvos(11.2)) {
    switch (subscriptionPeriod.unit) {
        case RCSubscriptionPeriodUnitDay:
            return @"DAY";
        case RCSubscriptionPeriodUnitWeek:
            return @"WEEK";
        case RCSubscriptionPeriodUnitMonth:
            return @"MONTH";
        case RCSubscriptionPeriodUnitYear:
            return @"YEAR";
    }
}

@end

File: ios/PurchasesHybridCommon/PurchasesHybridCommon/SKProduct+HybridAdditions.swift
//
//  SKProduct+HybridAdditions.swift
//  PurchasesHybridCommon
//
//  Created by Andrés Boedo on 4/13/22.
//  Copyright © 2022 RevenueCat. All rights reserved.
//

import Foundation
import StoreKit
import RevenueCat

@objc public extension StoreProduct {

    @objc var rc_dictionary: [String: Any] {
        var dictionary: [String: Any] = [
            "identifier": productIdentifier,
            "description": localizedDescription,
            "title": localizedTitle,
            "price": price,
            "price_string": localizedPriceString,
            "currency_code": currencyCode ?? NSNull(),
            "intro_price": NSNull(),
            "intro_price_string": NSNull(),
            "intro_price_period": NSNull(),
            "intro_price_period_unit": NSNull(),
            "intro_price_period_number_of_units": NSNull(),
            "intro_price_cycles": NSNull(),
            "introPrice": NSNull(),
            "discounts": NSNull(),
        ]

        if #available(iOS 11.2, tvOS 11.2, macOS 10.13.2, *),
           let introductoryDiscount = introductoryDiscount {
            dictionary["intro_price"] = introductoryDiscount.price
            dictionary["intro_price_string"] = introductoryDiscount.localizedPriceString
            dictionary["intro_price_period"] = StoreProduct.rc_normalized(subscriptionPeriod: introductoryDiscount.subscriptionPeriod)
            dictionary["intro_price_period_unit"] = StoreProduct.rc_normalized(subscriptionPeriodUnit: introductoryDiscount.subscriptionPeriod.unit)
            dictionary["intro_price_period_number_of_units"] = introductoryDiscount.subscriptionPeriod.value
            dictionary["intro_price_cycles"] = introductoryDiscount.numberOfPeriods
            dictionary["introPrice"] = introductoryDiscount.rc_dictionary
        }

        if #available(iOS 12.2, tvOS 12.2, macOS 10.14.4, *) {
            dictionary["discounts"] = discounts.map { $0.rc_dictionary }
        }

        return dictionary
    }

    @objc(rc_normalizedSubscriptionPeriod:)
    static func rc_normalized(subscriptionPeriod: SubscriptionPeriod) -> String {
        let unitString: String
        switch subscriptionPeriod.unit {
        case .day:
            unitString = "D"
        case .week:
            unitString = "W"
        case .month:
            unitString = "M"
        case .year:
            unitString = "Y"
        @unknown default:
            unitString = "-"
        }
        return "P\(subscriptionPeriod.value)\(unitString)"
    }

    @objc(rc_normalizedSubscriptionPeriodUnit:)
    static func rc_normalized(subscriptionPeriodUnit: SubscriptionPeriod.Unit) -> String {
        switch subscriptionPeriodUnit {
        case .day:
            return "DAY"
        case .week:
            return "WEEK"
        case .month:
            return "MONTH"
        case .year:
            return "YEAR"
        @unknown default:
            return "-"
        }
    }

}
